python/dgl/data/csv_dataset.py

```diff
@@ -0,0 +1,420 @@
+
+    @staticmethod
+    def load_edge_data_from_csv(meta_edge, data_parser, base_dir=None, separator=','):
+        df = CSVDataLoader.read_csv(meta_edge.file_name, base_dir, separator)
+        etype = tuple(meta_edge.etype)
+        edata = data_parser(df)
+        if meta_edge.src_id_field not in edata:
+            raise DGLError(f"src_id_field[{meta_edge.src_id_field}] does not exist in parsed edge data dict.")
+        if meta_edge.dst_id_field not in edata:
+            raise DGLError(f"dst_id_field[{meta_edge.dst_id_field}] does not exist in parsed edge data dict.")
+        src_ids = edata[meta_edge.src_id_field]
+        dst_ids = edata[meta_edge.dst_id_field]
+        graph_ids = edata[meta_edge.graph_id_field] if meta_edge.graph_id_field in edata else None
+        for key in [meta_edge.src_id_field, meta_edge.dst_id_field, meta_edge.graph_id_field]:
+            if key in edata:
+                edata.pop(key)
+        return EdgeData(src_ids, dst_ids, edata, type=etype, graph_id=graph_ids)
+
+    @staticmethod
+    def load_graph_data_from_csv(meta_graph, data_parser, base_dir=None, separator=','):
+        df = CSVDataLoader.read_csv(meta_graph.file_name, base_dir, separator)
+        gdata = data_parser(df)
+        if meta_graph.graph_id_field not in gdata:
+            raise DGLError(f"graph_id_field[{meta_graph.graph_id_field}] does not exist in parsed graph data dict.")
+        graph_ids = gdata[meta_graph.graph_id_field]
+        gdata.pop(meta_graph.graph_id_field)
+        return GraphData(graph_ids, gdata)
+
+
+class DGLGraphConstructor:
+    """ Class for constructing DGLGraph from Node/Edge/Graph data. Internal use only. """
+    @staticmethod
+    def construct_graphs(node_data, edge_data, graph_data=None):
+        if not isinstance(node_data, list):
+            node_data = [node_data]
+        if not isinstance(edge_data, list):
+            edge_data = [edge_data]
+        node_dict = DGLGraphConstructor._parse_node_data(node_data)
+        edge_dict = DGLGraphConstructor._parse_edge_data(edge_data, node_dict)
+        graph_dict = DGLGraphConstructor._construct_graphs(
+            node_dict, edge_dict)
+        if graph_data is None:
+            graph_data = GraphData(np.full(1, 0), {})
+        graphs, data = DGLGraphConstructor._parse_graph_data(
+            graph_data, graph_dict)
+        return graphs, data
+
+    @staticmethod
+    def _parse_node_data(node_data):
+        # node_ids could be arbitrary numeric values, namely non-sorted, duplicated, not labeled from 0 to num_nodes-1
+        node_dict = {}
+        for n_data in node_data:
+            graph_ids = np.unique(n_data.graph_id)
+            for graph_id in graph_ids:
+                if graph_id in node_dict and n_data.type in node_dict[graph_id]:
+                    raise DGLError(f"Duplicate node type[{n_data.type}] for same graph[{graph_id}], please place the same node_type for same graph into single NodeData.")
```

Comment:
I feel the check should not belong here. If there are multiple CSVs for the same node type, you should be able to detect that during YAML validation.

Reply:
Yes, we can detect in YAML validation, but this is where such duplicate takes side effects. as each `nodes.csv` are parsed into `NodeData` independently and appended into a list. and such logic is independent of YAML read/validate. Place the check here can also detect any unexpected processing during generate `NodeData`. I try to de-couple or de-compose each process stage from YAML to CSVDataset.

Reply:
The issue is the condition is hard to parse and the error message is also quite convoluted with little context. For example, what is NodeData? what is node_type? Is it a variable name or something else? I also don't understand what you mean by "unxpected processing". Could you elaborate?

Reply:
add `duplicate ntype/etype check` in `load_yaml_with_sanity_check`